### PR TITLE
CORDA-3608 Performance tune the new checkpoint schema

### PR DESCRIPTION
**Pull request by kyriathar**
Performance tuning of the new checkpoint schema;

- Checkpoint tables are now using `flowId` as join keys. 
- Indexes consist of a PK's index on `node_checkpoints(flow_id)` and then unique indexes on `node_checkpoint_blobs(flow_id)` and `node_flow_metadata(flow_id)`.
- Serialization of `checkpointState` is being done with `CHECKPOINT_CONTEXT` so that we can have compression. This is actually needed for whenever messages get passed into `checkpointState.sessions` therefore `checkpointState` grows in size upon serialized and saved into the database.

At the moment `DBFlowResult` and `DBFlowException` are not integrated. They will be integrated upon implementing [CORDA-3692](https://r3-cev.atlassian.net/browse/CORDA-3692) and [CORDA-3681](https://r3-cev.atlassian.net/browse/CORDA-3681).

### CODE REVIEW

**Comment by mnesbit** [comment]
Sorry, misunderstood the description, because you mean you have unique index on the flowID columns of the blob tables, not on the blobs themselves that doesn't work

File: node/src/main/resources/migration/node-core.changelog-v17.xml [hunk 1]
```diff
@@ -75,10 +66,10 @@
 
     <changeSet author="R3.Corda" id="add_new_flow_exceptions_table" dbms="!postgresql">
         <createTable tableName="node_flow_exceptions">
-            <column name="id" type="BIGINT">
+            <column name="flow_id" type="NVARCHAR(64)">
                 <constraints nullable="false"/>
             </column>
-            <column name="type" type="NVARCHAR(256)">
+            <column name="type" type="VARCHAR(255)">
```
Comment (mnesbit) [inline node/src/main/resources/migration/node-core.changelog-v17.xml]: Why is this one VARCHAR and all the others are NVARCHAR?

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 2]
```diff
@@ -273,36 +267,121 @@ class DBCheckpointStorage(
...
-                true
-            }
-        } ?: false
+        var deletedRows = 0
+        val flowId = id.uuid.toString()
+        deletedRows += deleteRow(DBFlowMetadata::class.java, DBFlowMetadata::flowId.name, flowId)
+        deletedRows += deleteRow(DBFlowCheckpointBlob::class.java, DBFlowCheckpointBlob::flowId.name, flowId)
+        deletedRows += deleteRow(DBFlowCheckpoint::class.java, DBFlowCheckpoint::flowId.name, flowId)
+//        resultId?.let { deletedRows += deleteRow(DBFlowResult::class.java, DBFlowResult::flow_id.name, it.toString()) }
+//        exceptionId?.let { deletedRows += deleteRow(DBFlowException::class.java, DBFlowException::flow_id.name, it.toString()) }
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:374]: commented out code

File: node/src/main/resources/migration/node-core.changelog-v17.xml (hunk 1 above)
Comment (kyriathar) [reply node/src/main/resources/migration/node-core.changelog-v17.xml]: Thanks for the catch, changed it to NVARCHAR.

**Review by lankydan — APPROVED** [review]
I'm happy with these. I'll let @rick-r3 give it a quick once over first.

**Review by rick-r3 — APPROVED** [review]

**Pull request merged**